src/Uno.UI/UI/Xaml/DependencyObjectStore.cs

```diff
@@ -1587,7 +1588,7 @@ private void PropagateInheritedNonLocalProperties(DependencyObjectStore? childSt
 			// properties.
 
 			// Ancestors is a local cache to avoid walking up the tree multiple times.
-			var ancestors = new AncestorsDictionary();
+			var ancestors = _ancestorsDictionary;
```

Comment:
I'm not sure this is enough. The problem here is that the hashtable may be cleared too early, causing ancestors to be cleaned in when children are processed. We may not have the appropriate amount of validation for that scenario either...

Reply:
The containing method turned out to be re-entrant.

So, using a singleton is a potential behavioral change that we don't have coverage for by the current tests.

However, the existing behavior might also be wrong if a parent could change after isAncestor is cached.

Reply:
It may be the case, but there are many manipulations that can happen when templates are applied/removed.

One thing that we could do is pool those dictionaries, as there may only be as many as the maximum depths of the current tree.

Reply:
Yup. Pooling makes most sense for now to at least preserve the original behavior. In future we may want to re-validate whether this code has a bug in certain scenarios.